<a href="https://colab.research.google.com/github/tavoAA05/Inteligencia_Artificial_SIS420/blob/main/laboratorios/lab1/Sis420_lab1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Laboratorio 1
**Materia:** SIS 420
**Estudiante:** Apaza Albarado Gustavo Gabriel     
**Descripción:** Programa en Python que encuentra los mejores valores de 'w' y 'b' para la ecuación y = b + wx utilizando la estrategia de Búsqueda en Rejilla (Grid Search) e infiere 10 nuevos valores.

In [ ]:
# 1. Dataset de Entrenamiento (Horas de estudio vs Calificación)
x_train = [1.0, 2.0, 3.0, 4.0, 5.0]
y_train = [3.4, 5.6, 7.5, 9.4, 11.6]

# 2. Variables para almacenar la mejor combinación
mejor_w = 0.0
mejor_b = 0.0
menor_error = float('inf')

# 3. Búsqueda en Rejilla (Grid Search)
paso = 0.01
limite_inferior = -500
limite_superior = 500

for iter_w in range(limite_inferior, limite_superior + 1):
    w_candidato = iter_w * paso
    for iter_b in range(limite_inferior, limite_superior + 1):
        b_candidato = iter_b * paso

        error_total = 0.0
        for i in range(len(x_train)):
            x_i = x_train[i]
            y_real = y_train[i]

            # Ecuación de la recta: y = b + wx
            y_pred = b_candidato + (w_candidato * x_i)

            # Cálculo del error absoluto manual
            diferencia = y_real - y_pred
            if diferencia < 0:
                diferencia = -diferencia

            error_total += diferencia

        # Evaluar si la combinación actual mejora el margen de error
        if error_total < menor_error:
            menor_error = error_total
            mejor_w = w_candidato
            mejor_b = b_candidato

print("=== RESULTADOS DEL ENTRENAMIENTO ===")
print(f"Mejor Pendiente (w): {mejor_w:.2f}")
print(f"Mejor Intersección (b): {mejor_b:.2f}")
print(f"Error Absoluto Mínimo: {menor_error:.4f}\n")

# 4. Inferencias para 10 nuevos valores de x
x_nuevos = [6.0, 7.0, 8.0, 9.0, 10.0, 11.0, 12.0, 13.0, 14.0, 15.0]

print("=== INFERENCIAS PARA NUEVOS DATOS ===")
print(f"{'x (Nuevos)':<12} | {'y_predicho (b + w*x)':<20}")
print("-" * 35)

for x_i in x_nuevos:
    y_predicho = mejor_b + (mejor_w * x_i)
    print(f"{x_i:<12.1f} | {y_predicho:<20.2f}")

=== RESULTADOS DEL ENTRENAMIENTO ===
Mejor Pendiente (w): 2.05
Mejor Intersección (b): 1.35
Error Absoluto Mínimo: 0.3000

=== INFERENCIAS PARA NUEVOS DATOS ===
x (Nuevos)   | y_predicho (b + w*x)
-----------------------------------
6.0          | 13.65               
7.0          | 15.70               
8.0          | 17.75               
9.0          | 19.80               
10.0         | 21.85               
11.0         | 23.90               
12.0         | 25.95               
13.0         | 28.00               
14.0         | 30.05               
15.0         | 32.10               


Resultados del Entrenamiento: Al finalizar la iteración sobre el espacio de parámetros, el programa arroja los siguientes valores óptimos:Mejor Pendiente ($w$): $2.05$ Mejor Intersección ($b$): $1.35$ Error Absoluto Mínimo: $0.3000$